hugr-passes/src/dataflow/results.rs

```diff
@@ -0,0 +1,137 @@
+
+/// Results of a dataflow analysis, packaged with the Hugr for easy inspection.
+/// Methods allow inspection, specifically [read_out_wire](Self::read_out_wire).
+pub struct AnalysisResults<V: AbstractValue, C: DFContext<V>> {
+    pub(super) ctx: C,
+    pub(super) in_wire_value: Vec<(Node, IncomingPort, PartialValue<V>)>,
+    pub(super) case_reachable: Vec<(Node, Node)>,
+    pub(super) bb_reachable: Vec<(Node, Node)>,
+    pub(super) out_wire_values: HashMap<Wire, PartialValue<V>>,
+}
+
+impl<V: AbstractValue, C: DFContext<V>> AnalysisResults<V, C> {
+    /// Allows to use the [HugrView] contained within
+    pub fn hugr(&self) -> &C::View {
+        &self.ctx
+    }
+
+    /// Gets the lattice value computed for the given wire
+    pub fn read_out_wire(&self, w: Wire) -> Option<PartialValue<V>> {
+        self.out_wire_values.get(&w).cloned()
+    }
+
+    /// Tells whether a [TailLoop] node can terminate, i.e. whether
+    /// `Break` and/or `Continue` tags may be returned by the nested DFG.
+    /// Returns `None` if the specified `node` is not a [TailLoop].
+    ///
+    /// [TailLoop]: hugr_core::ops::TailLoop
+    pub fn tail_loop_terminates(&self, node: Node) -> Option<TailLoopTermination> {
+        self.hugr().get_optype(node).as_tail_loop()?;
+        let [_, out] = self.hugr().get_io(node).unwrap();
+        Some(TailLoopTermination::from_control_value(
+            self.in_wire_value
+                .iter()
+                .find_map(|(n, p, v)| (*n == out && p.index() == 0).then_some(v))
+                .unwrap(),
+        ))
+    }
+
+    /// Tells whether a [Case] node is reachable, i.e. whether the predicate
+    /// to its parent [Conditional] may possibly have the tag corresponding to the [Case].
+    /// Returns `None` if the specified `case` is not a [Case], or is not within a [Conditional]
+    /// (e.g. a [Case]-rooted Hugr).
+    ///
+    /// [Case]: hugr_core::ops::Case
+    /// [Conditional]: hugr_core::ops::Conditional
+    pub fn case_reachable(&self, case: Node) -> Option<bool> {
+        self.hugr().get_optype(case).as_case()?;
+        let cond = self.hugr().get_parent(case)?;
+        self.hugr().get_optype(cond).as_conditional()?;
+        Some(
+            self.case_reachable
+                .iter()
+                .any(|(cond2, case2)| &cond == cond2 && &case == case2),
+        )
+    }
+
+    /// Tells us if a block ([DataflowBlock] or [ExitBlock]) in a [CFG] is known
+    /// to be reachable. (Returns `None` if argument is not a child of a CFG.)
+    ///
+    /// [CFG]: hugr_core::ops::CFG
+    /// [DataflowBlock]: hugr_core::ops::DataflowBlock
+    /// [ExitBlock]: hugr_core::ops::ExitBlock
+    pub fn bb_reachable(&self, bb: Node) -> Option<bool> {
+        let cfg = self.hugr().get_parent(bb)?; // Not really required...??
+        self.hugr().get_optype(cfg).as_cfg()?;
+        let t = self.hugr().get_optype(bb);
+        if !t.is_dataflow_block() && !t.is_exit_block() {
```

Comment:
I think this would be slightly more clear with an application de Morgan's law 🙂

Reply:
Done with `then_some`, how's that?